envoy/stats/symbol_table.h

```diff
@@ -1,235 +1,20 @@
-#include <utility>
-#include <vector>
-
-#include "envoy/common/pure.h"
-
-#include "absl/container/inlined_vector.h"
-#include "absl/strings/string_view.h"
-
 namespace Envoy {
 namespace Stats {
 
-/**
- * Runtime representation of an encoded stat name. This is predeclared only in
- * the interface without abstract methods, because (a) the underlying class
- * representation is common to both implementations of SymbolTable, and (b)
- * we do not want or need the overhead of a vptr per StatName. The common
- * declaration for StatName is in source/common/stats/symbol_table_impl.h
- */
-class StatName;
-using StatNameVec = absl::InlinedVector<StatName, 8>;
-
-class StatNameList;
-class StatNameSet;
-
-using StatNameSetPtr = std::unique_ptr<StatNameSet>;
+// Forward declarations for the symbol table classes. See
+// source/common/stats/symbol_table_impl.h" for the class definitions.
 
 /**
- * Holds a range of indexes indicating which parts of a stat-name are
- * dynamic. This is used to transfer stats from hot-restart parent to child,
- * retaining the same name structure.
+ * Runtime representation of an encoded stat name.
  */
-using DynamicSpan = std::pair<uint32_t, uint32_t>;
-using DynamicSpans = std::vector<DynamicSpan>;
+class StatName;
 
 /**
- * SymbolTable manages a namespace optimized for stat names, exploiting their
- * typical composition from "."-separated tokens, with a significant overlap
- * between the tokens. The interface is designed to balance optimal storage
- * at scale with hiding details from users. We seek to provide the most abstract
- * interface possible that avoids adding per-stat overhead or taking locks in
- * the hot path.
+ * Holds a set of symbols used to compose hierarhical names.
  */
-class SymbolTable {
-public:
-  /**
-   * Efficient byte-encoded storage of an array of tokens. The most common
-   * tokens are typically < 127, and are represented directly. tokens >= 128
-   * spill into the next byte, allowing for tokens of arbitrary numeric value to
-   * be stored. As long as the most common tokens are low-valued, the
-   * representation is space-efficient. This scheme is similar to UTF-8. The
-   * token ordering is dependent on the order in which stat-names are encoded
-   * into the SymbolTable, which will not be optimal, but in practice appears
-   * to be pretty good.
-   *
-   * This is exposed in the interface for the benefit of join(), which is
-   * used in the hot-path to append two stat-names into a temp without taking
-   * locks. This is used then in thread-local cache lookup, so that once warm,
-   * no locks are taken when looking up stats.
-   */
-  using Storage = uint8_t[];
-  using StoragePtr = std::unique_ptr<Storage>;
-
-  virtual ~SymbolTable() = default;
-
-  /**
-   * @return uint64_t the number of symbols in the symbol table.
-   */
-  virtual uint64_t numSymbols() const PURE;
-
-  /**
-   * Decodes a vector of symbols back into its period-delimited stat name. If
```

Comment:
It will be useful if all these comments are transferred to the header file in source.

Reply:
Yes that's a huge catch. Thank you. Done.

Also, doing that, I noticed some public APIs that can be private. There were still some hacky remnants of having a FakeSymbolTableImpl which shared some mechanism with SymbolTableImpl.